Obtain Location from appropriate Window directly, rather then Frame

sideshowbarker opened the pull request:
#### 5bf1af1ceb7a05c6f0bfc2bf4c890c0bcade1c7a
<pre>
Obtain Location from appropriate Window directly, rather then Frame
<a href="https://bugs.webkit.org/show_bug.cgi?id=261889">https://bugs.webkit.org/show_bug.cgi?id=261889</a>

Reviewed by Chris Dumez.

This change causes Location::url() to be obtained from the appropriate
Window instance directly, rather than a Frame.

Otherwise, without this change, the “document.open() does not change
document&apos;s URL (active but not fully active document)” test case from
<a href="https://wpt.fyi/results/html/webappapis/dynamic-markup-insertion/opening-the-input-stream/url.window.html">https://wpt.fyi/results/html/webappapis/dynamic-markup-insertion/opening-the-input-stream/url.window.html</a>
fails in WebKit, and a number of other tests time out unexpectedly.

Note: This change incidentally also corrects the expectations for the test at
<a href="https://wpt.fyi/results/html/webappapis/scripting/processing-model-2/unhandled-promise-rejections/disallow-crossorigin.html">https://wpt.fyi/results/html/webappapis/scripting/processing-model-2/unhandled-promise-rejections/disallow-crossorigin.html</a>
In its WebKit expectations file, that test been marked as a TIMEOUT —
but it’s marked as a FAIL in the wpt.fyi results.

* LayoutTests/http/tests/security/xss-DENIED-script-inject-into-inactive-window2-pson.html:
* LayoutTests/http/tests/security/xss-DENIED-script-inject-into-inactive-window2.html:
* LayoutTests/http/tests/security/xss-DENIED-script-inject-into-inactive-window3.html:
Adjusted expected location.href

* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/back-forward-cache/pushstate.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/back-forward-cache/service-worker-clients-claim.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/back-forward-cache/service-worker-clients-matchall.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/back-forward-cache/service-worker-controlled-after-restore.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/back-forward-cache/service-worker-unregister.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/back-forward-cache/storage-events-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/navigate-cross-origin-iframe-to-same-url-with-fragment-fire-load-event-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-history-interface/001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-history-interface/002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/no_window_open_when_term_nesting_level_nonzero.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/opening-the-input-stream/bailout-exception-vs-return-xml.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/opening-the-input-stream/ignore-opens-during-unload.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/opening-the-input-stream/url.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/scripting/event-loops/fully_active_document.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/scripting/processing-model-2/unhandled-promise-rejections/allow-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/scripting/processing-model-2/unhandled-promise-rejections/disallow-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/scripting/processing-model-2/unhandled-promise-rejections/promise-rejection-events-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/scripting/processing-model-2/unhandled-promise-rejections/promise-rejection-events-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/update-rendering/child-document-raf-order-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/user-prompts/cannot-show-simple-dialogs/confirm-different-origin-frame.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/user-prompts/cannot-show-simple-dialogs/prompt-different-origin-frame.sub-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/opening-the-input-stream/url.window-expected.txt:
* Source/WebCore/page/Location.cpp:
(WebCore::Location::url const):

Canonical link: <a href="https://commits.webkit.org/269235@main">https://commits.webkit.org/269235@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/156334190765db38b9c158bbd5a43df5fe9418df

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/21927 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/22147 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/22994 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/23809 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/20305 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/22172 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/26398 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/22459 "Built successfully") | [⏳ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/WPE-WK2-Tests-EWS "Waiting to run tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/22157 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/21793 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/19018 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/24661 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/18923 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/19872 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/26136 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/19951 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/20094 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/23999 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/20528 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/17488 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/19877 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/19681 "Passed tests") | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5237 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/24084 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/20478 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/page/Location.cpp:
```diff
@@ -50,10 +51,7 @@ Location::Location(LocalDOMWindow& window)
-        return aboutBlankURL();
-
-    const URL& url = frame()->document()->urlForBindings();
+    const URL& url = LocalDOMWindowProperty::window()->document()->urlForBindings();
```
Can we use `m_window->document()` here and null check `m_window` instead of `frame()` above?

sideshowbarker replied:
> Can we use `m_window->document()` here and null check `m_window` instead of `frame()` above?

Yes, will make that change and push an update

sideshowbarker replied:
So compiling `m_window->document()` fails with _“error: no member named 'document' in 'WebCore::DOMWindow'”_.

And `LocalDOMWindowProperty::window()->document()` _also_ now fails with _“error: call to non-static member function without an object argument”_ — due to the changes made in 194c42fd694410278bc8a19d998239cfe043667f.

I guess I need to study the 194c42fd694410278bc8a19d998239cfe043667f changes to figure out what would actually work here instead now.
cc @achristensen07

cdumez replied:
Looks like Alex currently doesn't deal property with remote frames for Location::url(), as per the code:
```
const URL& Location::url() const
{
    RefPtr localFrame = dynamicDowncast<LocalFrame>(frame());
    if (!localFrame)
        return aboutBlankURL();
```

As a result, for now, I'd just do:
```
if (is<LocalDOMWindow>(m_domWindow.get())
    return aboutBlankURL();

const URL& url = downcast<LocalDOMWindow>(*m_domWindow).document()->urlForBindings();
```

sideshowbarker replied:
OK I can build successfully with `const URL& url = downcast<LocalDOMWindow>(*m_window).document()->urlForBindings()`, but the build doesn’t pass the _“does not change document's URL (active but not fully active document)”_ test at http://wpt.live/html/webappapis/dynamic-markup-insertion/opening-the-input-stream/url.window.html

sideshowbarker replied:
OK, with `const URL& url = downcast<LocalDOMWindow>(*m_window).document()->urlForBindings()` it passes the http://wpt.live/html/webappapis/dynamic-markup-insertion/opening-the-input-stream/url.window.html test once again — if I also completely (re)remove the preceding null-frame check:

```c++
RefPtr localFrame = dynamicDowncast<LocalFrame>(frame());
if (!localFrame)
    return aboutBlankURL();
```
I believe that check isn’t actually spec-conformant — which is why the patch in this PR branch had also removed the previous (`s/if (!localFrame)/if (!frame())`) formulation of it:

https://github.com/WebKit/WebKit/pull/18026/commits/cfa36aa83397066f2721e40e6a2cf77053619a89#diff-9fb3dfaab8c435ab08296c7a653f8f5386788616337f1fc2004a8c6e5db3e276L53

cdumez commented on Source/WebCore/page/Location.cpp:
```diff
@@ -64,11 +65,10 @@ const Frame* Location::frame() const
-    RefPtr localFrame = dynamicDowncast<LocalFrame>(frame());
-    if (!localFrame)
-        return aboutBlankURL();
+    ASSERT(m_window);
```
Please add a null check for m_window, as suggested earlier.

cdumez commented on Source/WebCore/page/Location.cpp:
```diff
@@ -64,11 +65,10 @@ const Frame* Location::frame() const
-    if (!localFrame)
-        return aboutBlankURL();
+    ASSERT(m_window);
+    ASSERT(downcast<LocalDOMWindow>(*m_window).document());
```
This downcast isn't safe, as far as I know, this can be a RemoteDOMWindow as of recently. It should be a dynamicDowncast as suggested earlier.

sideshowbarker replied:
OK, now a dynamicDowncast

cdumez reviewed (changes requested):
Needs changes

cdumez commented on Source/WebCore/page/Location.cpp:
```diff
@@ -64,11 +65,12 @@ const Frame* Location::frame() const
-    RefPtr localFrame = dynamicDowncast<LocalFrame>(frame());
-    if (!localFrame)
-        return aboutBlankURL();
+    RefPtr localWindow = dynamicDowncast<LocalDOMWindow>(*m_window);
```
Still not safe, here is what it should be:
```
RefPtr localWindow = dynamicDowncast<LocalDOMWindow>(m_window.get());
if (!localWindow || !localWindow->document())
    return aboutBlankURL();
```

cdumez commented on Source/WebCore/page/Location.cpp:
```diff
@@ -64,11 +65,11 @@ const Frame* Location::frame() const
-        return aboutBlankURL();
+    RefPtr localWindow = dynamicDowncast<LocalDOMWindow>(*m_window);
+    if (!localWindow)
+        return *new URL(emptyString());
```
should be `return aboutBlankURL();`

cdumez replied:
You're leaking memory here.

sideshowbarker replied:
> should be `return aboutBlankURL();`

As mentioned at https://github.com/WebKit/WebKit/pull/18026#discussion_r1354620288, returning `aboutBlankURL()` if the window (or frame) is null isn’t spec-conformant — it causes the _“document.open() does not change document's URL (active but not fully active document)”_ test case at
https://wpt.fyi/results/html/webappapis/dynamic-markup-insertion/opening-the-input-stream/url.window.html
to fail. So it’s unclear to me what otherwise we can return here.

cdumez replied:
Then something like this:
```
if (!localWindow) {
    static NeverDestroyed<URL> emptyURL = emptyString();
    return emptyURL.get();
}
```

Your current proposal is leaking memory, which is what I am objecting to.

If this works, this would be better too:
```
if (!localWindow) {
    static NeverDestroyed<URL> nullURL;
    return nullURL.get();
}
```

sideshowbarker replied:
> If this works, this would be better too:
> 
> ```
> if (!localWindow) {
>     static NeverDestroyed<URL> nullURL;
>     return nullURL.get();
> }
> ```

Thanks — that works. Pushed now now

cdumez reviewed (approved):
LGTM

sideshowbarker committed:
Obtain Location from appropriate Window directly, rather then Frame
https://bugs.webkit.org/show_bug.cgi?id=261889

Reviewed by Chris Dumez.

This change causes Location::url() to be obtained from the appropriate
Window instance directly, rather than a Frame.

Otherwise, without this change, the “document.open() does not change
document's URL (active but not fully active document)” test case from
https://wpt.fyi/results/html/webappapis/dynamic-markup-insertion/opening-the-input-stream/url.window.html
fails in WebKit, and a number of other tests time out unexpectedly.

Note: This change incidentally also corrects the expectations for the test at
https://wpt.fyi/results/html/webappapis/scripting/processing-model-2/unhandled-promise-rejections/disallow-crossorigin.html
In its WebKit expectations file, that test been marked as a TIMEOUT —
but it’s marked as a FAIL in the wpt.fyi results.

* LayoutTests/http/tests/security/xss-DENIED-script-inject-into-inactive-window2-pson.html:
* LayoutTests/http/tests/security/xss-DENIED-script-inject-into-inactive-window2.html:
* LayoutTests/http/tests/security/xss-DENIED-script-inject-into-inactive-window3.html:
Adjusted expected location.href

* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/back-forward-cache/pushstate.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/back-forward-cache/service-worker-clients-claim.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/back-forward-cache/service-worker-clients-matchall.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/back-forward-cache/service-worker-controlled-after-restore.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/back-forward-cache/service-worker-unregister.https-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/back-forward-cache/storage-events-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/browsing-the-web/navigating-across-documents/navigate-cross-origin-iframe-to-same-url-with-fragment-fire-load-event-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-history-interface/001-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/history/the-history-interface/002-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/browsers/the-window-object/open-close/no_window_open_when_term_nesting_level_nonzero.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/opening-the-input-stream/bailout-exception-vs-return-xml.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/opening-the-input-stream/ignore-opens-during-unload.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/opening-the-input-stream/url.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/scripting/event-loops/fully_active_document.window-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/scripting/processing-model-2/unhandled-promise-rejections/allow-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/scripting/processing-model-2/unhandled-promise-rejections/disallow-crossorigin-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/scripting/processing-model-2/unhandled-promise-rejections/promise-rejection-events-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/scripting/processing-model-2/unhandled-promise-rejections/promise-rejection-events-iframe-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/update-rendering/child-document-raf-order-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/user-prompts/cannot-show-simple-dialogs/confirm-different-origin-frame.sub-expected.txt:
* LayoutTests/imported/w3c/web-platform-tests/html/webappapis/user-prompts/cannot-show-simple-dialogs/prompt-different-origin-frame.sub-expected.txt:
* LayoutTests/platform/glib/imported/w3c/web-platform-tests/html/webappapis/dynamic-markup-insertion/opening-the-input-stream/url.window-expected.txt:
* Source/WebCore/page/Location.cpp:
(WebCore::Location::url const):

Canonical link: https://commits.webkit.org/269235@main
